scrapper/src/main/java/edu/java/domain/JDBCLinkDao.java

```diff
@@ -0,0 +1,141 @@
+        var id =
+            jdbcTemplate.queryForObject(SQL, (rs, rowNum) -> rs.getLong("id"), lastUpdate, link.getId(), link.getId());
+        String SQL1 = "select * from link where id=?";
+        return jdbcTemplate.queryForObject(SQL1, (rs, rowNum) -> new Link(
+            rs.getLong("id"),
+            URI.create(rs.getString("url")),
+            rs.getString("description"),
+            rs.getTimestamp("last_updated").toInstant().atOffset(ZoneOffset.UTC)
+        ), id);
+    }
+
+    public Link deleteLink(long chatId, URI url) {
+        String SQL1 = "select * from link where url = ?";
+        var link = jdbcTemplate.queryForObject(SQL1, (rs, rowNum) -> new Link(
+            rs.getLong("id"),
+            URI.create(rs.getString("url")),
+            rs.getString("description"),
+            rs.getTimestamp("last_updated").toInstant().atOffset(ZoneOffset.UTC)
+        ), url.toString());
+        String SQL =
+            "with d as (delete from public.assignment where assignment.link_id = ?) delete from link where url=?";
+        assert link != null;
+        jdbcTemplate.update(SQL, link.getId(), url.toString());
+        return link;
+    }
+
+    public List<Link> findAllLinksWithLastUpdateEarlierThan(OffsetDateTime lastUpdate) {
+        String SQL = "select  * from link where last_updated < ?";
+        return jdbcTemplate.query(SQL, (rs, rowNum) -> new Link(
+            rs.getLong("id"),
+            URI.create(rs.getString("url")),
+            rs.getString("description"),
+            rs.getTimestamp("last_updated").toInstant().atOffset(ZoneOffset.UTC)
+        ), lastUpdate);
+    }
+
+    public List<Link> findAllLinks() {
+        String SQL = "select  * from link";
+        return jdbcTemplate.query(SQL, (rs, rowNum) -> new Link(
+            rs.getLong("id"),
+            URI.create(rs.getString("url")),
+            rs.getString("description"),
+            rs.getTimestamp("last_updated").toInstant().atOffset(ZoneOffset.UTC)
+        ));
+    }
+
+    public long findByUrlAndChat(long chatId, URI url) {
+        String SQL = "select link.id from link where url = ?";
+        var linkId = jdbcTemplate.queryForObject(SQL, (rs, rowNum) ->
+                rs.getLong("id"),
+            url.toString()
+        );
+        String SQL2 = "select assignment.chat_id from assignment where link_id = ?";
+        var chatIds = jdbcTemplate.query(SQL2, (rs, rowNum) ->
+                rs.getLong("chat_id"),
+            linkId
+        );
+        if (chatIds.contains(chatId)) {
+            return chatId;
+        } else {
+            return -1L;
+        }
+    }
+
+    public List<Link> findAllLinksFromChat(long chatId) {
+        String SQL1 = "select assignment.link_id from assignment where chat_id = ?";
+        List<Long> ids = jdbcTemplate.query(SQL1, (rs, rowNum) ->
+                rs.getLong("link_id"),
+            chatId
+        );
+        List<Link> links = new ArrayList<>();
+        if (!ids.isEmpty()) {
```

Comment:
я бы сделал revert ифа
```suggestion
        if (ids.isEmpty()) return List.of()
```

Reply:
да в курсе про один из принципов clean code. Но не люблю опаздывать, поэтому делала как быстрее.